app/server/lib/Assistance.ts

```diff
@@ -222,27 +253,33 @@ export class OpenAIAssistant implements Assistant {
   }
 
   private async _fetchCompletion(messages: AssistanceMessage[], userIdHash: string, longerContext: boolean) {
+    const model = longerContext ? this._longerContextModel : this._model;
     const apiResponse = await DEPS.fetch(
       this._endpoint,
       {
         method: "POST",
         headers: {
-          "Authorization": `Bearer ${this._apiKey}`,
+          ...(this._apiKey ? {
+            "Authorization": `Bearer ${this._apiKey}`,
+          } : undefined),
           "Content-Type": "application/json",
         },
         body: JSON.stringify({
           messages,
           temperature: 0,
-          model: longerContext ? OpenAIAssistant.LONGER_CONTEXT_MODEL : OpenAIAssistant.DEFAULT_MODEL,
+          ...(model ? { model } : undefined),
           user: userIdHash,
+          ...(OpenAIAssistant.MAX_TOKENS ? {
+            max_tokens: OpenAIAssistant.MAX_TOKENS,
+          } : undefined),
         }),
       },
     );
     const resultText = await apiResponse.text();
     const result = JSON.parse(resultText);
     const errorCode = result.error?.code;
     if (errorCode === "context_length_exceeded" || result.choices?.[0].finish_reason === "length") {
-      if (!longerContext) {
+      if (!longerContext && this._longerContextModel) {
```

Comment:
I'm surprised in general by the implication that the non-openai endpoints are compatible with the same client code. This is probably the place where that applies the most. Do you actually have models with different context lengths? Do they return `"error": {"code": "context_length_exceeded", ...}` or `"choices": [{"finish_reason": "length", ...`?

Reply:
Here, the context length fallback code is an obstacle I'm trying to overcome, rather than a feature I'm trying to use. I'm shooting for:
 * grist-saas behavior stays unchanged.
 * setting `ASSISTANT_CHAT_COMPLETION_ENDPOINT` makes a single request.

The logic in this PR is just me trying to thread the needle as cleanly as I can.

That said, the `python-llama-cpp` server was delightly faithful, I'm pretty sure I had to deal with this because the appropriate error code was being returned (but it's been a while since I actually dealt with that bit so I won't swear to it).